components/locid/src/parser/mod.rs

```diff
@@ -7,9 +7,82 @@ mod langid;
 mod locale;
 
 pub use errors::ParserError;
-pub use langid::{parse_language_identifier, parse_language_identifier_from_iter, ParserMode};
+pub use langid::{
+    parse_language_identifier, parse_language_identifier_from_iter,
+    parse_language_identifier_without_variants, ParserMode,
+};
 pub use locale::parse_locale;
 
-pub fn get_subtag_iterator(t: &[u8]) -> impl Iterator<Item = &[u8]> {
-    t.split(|c| *c == b'-' || *c == b'_')
+pub const fn get_subtag_iterator(slice: &[u8]) -> SubtagIterator {
+    let mut current_start = 0;
+    while current_start < slice.len()
+        && (slice[current_start] == b'_' || slice[current_start] == b'-')
+    {
+        current_start += 1;
+    }
+    let mut current_end = current_start;
+    while current_end < slice.len() && slice[current_end] != b'_' && slice[current_end] != b'-' {
+        current_end += 1;
+    }
+    SubtagIterator {
+        slice,
+        current_start,
+        current_end,
+    }
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct SubtagIterator<'a> {
+    slice: &'a [u8],
+    current_start: usize,
+    current_end: usize,
+}
+
+pub type ManualSlice<'a> = (&'a [u8], usize, usize);
+
+impl<'a> SubtagIterator<'a> {
+    pub const fn next_manual(mut self) -> (Self, Option<ManualSlice<'a>>) {
+        if self.current_start == self.current_end {
+            (self, None)
+        } else {
+            let r = (self.slice, self.current_start, self.current_end);
+            self.current_start = self.current_end;
+            while self.current_start < self.slice.len()
+                && (self.slice[self.current_start] == b'_'
+                    || self.slice[self.current_start] == b'-')
```

Comment:
`nit`: reverse the order (and consider `['-', '_'].contains` - optimize for one more lookup on the happy path.

Reply:
`contains` isn't `const`